alloc/src/lib.rs

```diff
@@ -0,0 +1,112 @@
+//! The simplest thing resembling an "allocator" I could possibly create.
+//! Allocates into a "large" static array.
+#![no_std]
+
+use core::alloc::Layout;
+use core::cell::UnsafeCell;
+use core::mem::MaybeUninit;
+use core::ptr;
+use core::sync::atomic::{AtomicUsize, Ordering};
+
+#[cfg(target_os = "none")]
+use core::alloc::GlobalAlloc;
+
+// 640k is enough for anyone
+const HEAP_SIZE: usize = 640 * 1024;
+
+#[repr(align(16))]
+struct Heap(UnsafeCell<MaybeUninit<[u8; HEAP_SIZE]>>);
+
+unsafe impl Sync for Heap {}
+
+static HEAP: Heap = Heap(UnsafeCell::new(MaybeUninit::uninit()));
+static USED: AtomicUsize = AtomicUsize::new(0);
+
+/// NOTABLE INVARIANTS:
+///  * `Layout` is non-zero sized (enforced by `GlobalAlloc`)
+///  * `align` is a power of two (enforced by `Layout::from_size_align`)
+pub unsafe fn alloc(layout: Layout) -> *mut u8 {
+    let heap = HEAP.0.get() as *mut u8;
+
+    let mut prev = USED.load(Ordering::Relaxed);
+    loop {
+        let align_offset = heap.add(prev).align_offset(layout.align());
+
+        // NOTE: layout.size() is non-zero. The check will fail if
+        // `align_offset` exceeds `HEAP_SIZE - prev`.
+        let space = (HEAP_SIZE - prev).saturating_sub(align_offset);
+        if space < layout.size() {
+            return ptr::null_mut();
+        }
+
+        let next = prev + align_offset + layout.size();
+        match USED.compare_exchange_weak(prev, next, Ordering::Relaxed, Ordering::Relaxed) {
```

Comment:
"Relaxed, Relaxed" kinda scares me — at a glance, it seems like there should be an acquire/release pair guarding the actual handing out of ptrs, which mustn't be reordered?

Reply:
I might be wrong, but I _think_ it's OK because we never free memory.
There are no writes to other memory regions than the atomic which need to be made visible to other allocating threads, as each allocation region is fully disjoint, so we just need to ensure that reads and writes to the `USED` atomic are consistent, which `Relaxed` handles.

Reply:
ah, hmm, yeah i think you're right!

Reply:
ah, hmm, yeah i think you're right!

Reply:
also it doesn't actually matter, because by the time we know about "cores" we will probably have switched over to an allocation strategy that supports "deallocating"

Reply:
"deallocating"?! never heard of it :stuck_out_tongue: